src/operator/nn/softmax-inl.h

```diff
@@ -464,35 +755,56 @@ void SoftmaxGradCompute(const nnvm::NodeAttrs& attrs,
   mxnet::TShape shape = AxisShapeCompact(inputs[0].shape_, &axis, true);
 
   int out_idx = softmax_has_dtype_override(attrs) ? 2 : 1;
+  out_idx = softmax_use_length(attrs) ? 3 : out_idx;
   bool safe_acc = dmlc::GetEnv("MXNET_SAFE_ACCUMULATION", false);
 
   MXNET_REAL_ACC_TYPE_SWITCH(inputs[0].type_flag_, OType, AType, {
     MSHADOW_REAL_TYPE_SWITCH(outputs[0].type_flag_, DType, {
       MXNET_ASSIGN_REQ_SWITCH(req[0], Req, {
-        if (safe_acc) {
-          if (shape.ndim() == 2) {
-            SoftmaxGrad<OP1, OP2, Req, negate, AType>(
-                ctx.get_stream<xpu>(), inputs[out_idx].dptr<OType>(),
-                inputs[0].dptr<OType>(), outputs[0].dptr<DType>(),
-                shape.get<2>(), axis, static_cast<DType>(temperature));
+        if (!softmax_use_length(attrs)) {
+          if (safe_acc) {
+            if (shape.ndim() == 2) {
+              SoftmaxGrad<OP1, OP2, Req, negate, AType>(
+                  ctx.get_stream<xpu>(), inputs[out_idx].dptr<OType>(),
+                  inputs[0].dptr<OType>(), outputs[0].dptr<DType>(),
+                  shape.get<2>(), axis, static_cast<DType>(temperature));
+            } else {
+              SoftmaxGrad<OP1, OP2, Req, negate, AType>(
+                  ctx.get_stream<xpu>(), inputs[out_idx].dptr<OType>(),
+                  inputs[0].dptr<OType>(), outputs[0].dptr<DType>(),
+                  shape.get<3>(), axis, static_cast<DType>(temperature));
+            }
           } else {
-            SoftmaxGrad<OP1, OP2, Req, negate, AType>(
-                ctx.get_stream<xpu>(), inputs[out_idx].dptr<OType>(),
-                inputs[0].dptr<OType>(), outputs[0].dptr<DType>(),
-                shape.get<3>(), axis, static_cast<DType>(temperature));
+            if (shape.ndim() == 2) {
+              SoftmaxGrad<OP1, OP2, Req, negate, DType>(
+                  ctx.get_stream<xpu>(), inputs[out_idx].dptr<OType>(),
+                  inputs[0].dptr<OType>(), outputs[0].dptr<DType>(),
+                  shape.get<2>(), axis, static_cast<DType>(temperature));
+            } else {
+              SoftmaxGrad<OP1, OP2, Req, negate, DType>(
+                  ctx.get_stream<xpu>(), inputs[out_idx].dptr<OType>(),
+                  inputs[0].dptr<OType>(), outputs[0].dptr<DType>(),
+                  shape.get<3>(), axis, static_cast<DType>(temperature));
+            }
           }
         } else {
-          if (shape.ndim() == 2) {
-            SoftmaxGrad<OP1, OP2, Req, negate, DType>(
+          MXNET_INT_TYPE_SWITCH(inputs[2].type_flag_, IType, {
```

Comment:
Do we really need to iterate different int types for length? Can we just cast the type to int64_t.

Reply:
It may not necessarily be int64_t, and what do you mean by "cast the type to int64_t"? Allocating a new buffer on the fly within the operator to hold the casted `length` input? I would consider that a big performance bottleneck.

Reply:
It does not necessarily be `int64_t`. I think you can use the same data type as M, which is `index_t` I think.

Reply:
I understand length is a tensor here. I guess my question is do we really need to care what the dtype of the values from length tensor and use a  `MXNET_INT_TYPE_SWITCH` macro here to iterate. Can we just simply cast them to `index_t` regardless of the dtype of `length` tensor.

Reply:
1. There's no iteration at all, `TYPE_SWITCH` is always a switch, not a loop. I would suggest that you read the code more carefully.
2. No matter you're doing a cast to whatever type within the kernel or not will not get rid of the `TYPE_SWITCH` at all. One way of doing a constant cast to `index_t` is within the computation kernel, like:
```c++
template<typename IType>
MSHADOW_XINLINE static void Map(int i, IType *buffer) {
  index_t val = static_cast<index_t>(buffer[i]);
}
```
And this would still require a `TYPE_SWITCH` when you're launching the kernel since this kernel is still templated and the input type is not limited.
Or if you really hate the `TYPE_SWITCH` for the compute kernel, you then have to cast your input buffer first:
```c++
Tensor<xpu, 1, index_t> index_t_buffer = ctx.requested[0].get_space_typed<xpu, 1, index_t>(<some shape>);
// Cast your buffer to index_t_buffer
// Launch the compute kernel with index_t_buffer, now you only need one TYPE_SWITCH for your input data type.
```
This I would really hate cause it's not bringing any benefit at all.
Or you only support `length` buffer of type `int64`, then you still need to use `MSHADOW_IDX_TYPE_SWITCH` to provide the error message when `length` is of other data types. Or you could insert an additional check on the data type before the kernel launches, but that would drop support for other integer types.
3. The fact that `length` is a tensor means that `length=M` is not an option, maybe you want to say `length=nullptr`? That would cause additional if branches within the kernel and will especially impact performance on GPU. The way this feature is implemented now puts that if branch on the host rather than within the kernel, which will have minimal performance impact.

Reply:
Please see my suggestion in the code. Can we just cast length[i] to index_t?

Reply:
Your original comment was "Do we really need to iterate different int types for length? Can we just cast the type to int64_t." and now you're essentially requesting something else in the code.
So regarding your new request, what's the point of casting an integer of a smaller data type to `index_t` if we're using it's original value only?

Reply:
My original comment was meant for checking different int types. "iterate" may not be an accurate word here. But nonetheless the suggestion follows the same line. I.e. only use one fixed type instead of various int types.

Reply:
I think your communication was not clear enough in the first place, I've now added cast to `index_t` in some existing code too, are you now comfortable with the changes now?

Reply:
Thanks for making the changes. Sorry, if my earlier comments caused you confusion.

Looking at the current code, I see the `SoftmaxWithLength()` and `Softmax()` are very similar now. (differ by one if block https://github.com/apache/incubator-mxnet/pull/15169/files#diff-be02d7c5660bf5cd623601a501fc7abeR134). Do you think we can combine this two functions into one now? :)

Reply:
No, because your are only looking at the CPU version of the code.

Reply:
Are you referring to these two kernel functions?
https://github.com/apache/incubator-mxnet/blob/50c1c7ca7f2e6a7864e7a0aedc775ae7dd8be091/src/operator/nn/softmax-inl.h#L283
https://github.com/apache/incubator-mxnet/blob/50c1c7ca7f2e6a7864e7a0aedc775ae7dd8be091/src/operator/nn/softmax-inl.h#L339

They also only differ by a few lines. Would you think a better way to consolidate them?

Reply:
I have a separate PR  #15545 to optimize softmax GPU implementation, I can look into merging those 2 kernels in that PR.